pkg/backends/cloudwatch/cloudwatch.go

```diff
@@ -170,58 +198,128 @@ func (client Client) buildMetricData(metrics *gostatsd.MetricMap) (metricData []
 	return metricData
 }
 
-// SendMetricsAsync sends the metrics in a MetricsMap to AWS Cloudwatch,
-// preparing payload synchronously but doing the send asynchronously.
-func (client Client) SendMetricsAsync(ctx context.Context, metrics *gostatsd.MetricMap, cb gostatsd.SendCallback) {
-	api := client.cloudwatch
-	metricData := client.buildMetricData(metrics)
-	length := len(metricData)
-	errors := []error{}
+// setupFlushTimer will either increase the timeout on the flush timer, or create the flush timer if it doesn't
+// exist yet.  It returns true or false to indicate whether or not the caller is responsible for servicing the
+// timer when it actually expires (true if it should).  Thread safe.
+func (client *Client) setupFlushTimer() bool {
+	client.timerMutex.Lock()
+	defer client.timerMutex.Unlock()
+	if client.timer != nil {
+		client.timer.Reset(client.batchDuration)
+		return false
+	} else {
+		client.timer = time.NewTimer(client.batchDuration)
+		return true
+	}
+}
 
-	if length < 1 {
-		cb(errors)
+// clearFlushTimer will stop the flush timer.  Thread safe.
+func (client *Client) clearFlushTimer() {
+	client.timerMutex.Lock()
+	client.timer = nil // Stop current timer
+	client.timerMutex.Unlock()
+}
+
+func (client *Client) flush() {
+	if !client.setupFlushTimer() {
 		return
 	}
 
-	go func() {
-		start := 0
+	// This is safe to read outside the timerMutex, as only the
+	// goroutine that created the initial timer can ever be here
+	<-client.timer.C
```

Comment:
I don't believe so:

Only one goroutine (call it "A") can be on line 230 at a time, and it's the goroutine that wrote to the timer. Essentially the flush timer is a mutex - if it's present, then the thing that created it will reach here, otherwise it won't. At line 235 the "flush timer as a mutex" is released.

I've called out in my comment above `clearFlushTimer` that it's possible for another goroutine (call it "B") to come along and take the data before "A" does, however that is protected by `queueMutex`, and just results in "A" having nothing to do.

That being said - I refactored this area specifically to make it more clear.  Even if it's safe, if it's still not clear, then I need to make it clearer. A reader of the code shouldn't have to think it through in that detail.

If you can confirm or refute my logic, I'll see about making it easier to follow / fixing it.

Reply:
I didn't have enough time to properly check the code here. It's probably fine if you say so :)